org/mozilla/jss/ssl/javax/buffer.c

```diff
@@ -0,0 +1,261 @@
+#include <stdbool.h>
+#include <string.h>
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "buffer.h"
+
+j_buffer *jb_alloc(size_t length) {
+    j_buffer *buf = calloc(1, sizeof(j_buffer));
+    buf->contents = calloc(length, sizeof(uint8_t));
+
+    buf->capacity = length;
+
+    // In the beginning, we can only write, not read. Hence, set our read_pos
+    // to the sentinel value, buf->capacity.
+    buf->write_pos = 0;
+    buf->read_pos = length;
+
+    return buf;
+}
+
+size_t jb_capacity(j_buffer *buf) {
```

Comment:
It would make sense to provide a remaining capacity function, but I'll refrain for now. I'm not sure if this'll be necessary for SSLEngine support; I think it could make it cleaner.